bulldozer/signals.go

```diff
@@ -21,62 +21,186 @@ import (
 	"strings"
 
 	"github.com/palantir/bulldozer/pull"
+	"github.com/pkg/errors"
 	"github.com/rs/zerolog"
 )
 
+type Signal interface {
+	Enabled() bool
+	// Determine if the signal has values assigned to it and should be considered when matching
+
+	Matches(context.Context, pull.Context, string) (bool, string, error)
+	// Determine if the signal matches a value in the target pull request
+}
+
+type Labels []string
+type CommentSubstrings []string
+type Comments []string
+type PRBodySubstrings []string
+type Branches []string
+type BranchPatterns []string
+type MaxCommits int
+
 type Signals struct {
-	Labels            []string `yaml:"labels"`
-	CommentSubstrings []string `yaml:"comment_substrings"`
-	Comments          []string `yaml:"comments"`
-	PRBodySubstrings  []string `yaml:"pr_body_substrings"`
-	Branches          []string `yaml:"branches"`
-	BranchPatterns    []string `yaml:"branch_patterns"`
+	Labels            `yaml:"labels"`
+	CommentSubstrings `yaml:"comment_substrings"`
+	Comments          `yaml:"comments"`
+	PRBodySubstrings  `yaml:"pr_body_substrings"`
+	Branches          `yaml:"branches"`
+	BranchPatterns    `yaml:"branch_patterns"`
+	MaxCommits        `yaml:"max_commits"`
```

Comment:
While a bit more verbose, I'd make these named fields instead of embedded structs for clarity, since that is how they are used, e.g.:

```go
Labels Labels `yaml:"labels"`
```

Reply:
Ok, will do. Perhaps it would be helpful to add a `Signal` suffix to each type to differentiate the name and the type?

Reply:
Yeah, if you want to, that sounds fine - I don't have a strong opinion either way.